File: apps/ios/WindmillKit/Sources/WindmillGym/StatisticsScreen.swift
import SwiftUI
import WindmillPlatform

// STATISTICS — the long window, and the second place a lifter can stand in this room at rest. It
// RENDERS what `Stats.board` decided and computes nothing itself: no e1RM, no normalisation, no
// sorting, no arithmetic in a view body.
//
// A CHART ON A PHONE HAS NO SCRUB AND NO HOVER, so every number a mark would otherwise reveal on
// touch is printed beside it — each bar row carries its own peak, each movement line its two
// endpoints. A chart whose values can only be reached by a gesture the surface does not have is a
// decoration, and this product does not draw those.
//
// Nothing here is a grade. No score, no percentage, no green and no red: iris is the one accent
// and it reads as iron, exactly as it does on the weight the lifter is under.

struct StatisticsScreen: View {
    @ObservedObject var store: TrainingStore

    @Environment(\.gymSkin) private var skin
    @State private var board: Stats.Board?
    @State private var failure: TrainingStore.WriteFailure?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: WindmillSpace.x5) {
                head
                if let board {
                    if let weeks = board.weeks { self.weeks(weeks) }
                    ForEach(board.lines) { line in movement(line) }
                    if board.isEmpty { nothingYet }
                } else if let failure {
                    // In the log's own words when it sent any. A 500 that said "internal error" is
                    // not a phone with no signal, and this screen may not report it as one.
                    silence(failure.line("these lines aren’t drawn"), retry: true)
                } else {
                    silence("reading your log…", retry: false)
                }
            }
            .padding(.horizontal, WindmillSpace.x5)
            .padding(.top, WindmillSpace.x10)
            .padding(.bottom, WindmillSpace.x8)
        }
        // Read on the way in and shaped once, outside the body. There is no cache to key on a
        // collection's length and get wrong: a second visit asks the log again, and a set logged in
        // between counts the next time this screen is opened.
        .task { await read() }
    }

    private var head: some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x1) {
            Text("Statistics")
                .font(WindmillFont.display(32))
                .foregroundStyle(skin.ink)
            Text("every session you have finished")
                .font(GymType.numeral(13))
                .foregroundStyle(skin.inkFaint)
        }
    }

    // Two rows of bars, each measured against ITS OWN peak — sessions and working sets count
    // different things, and one axis across both is the chart that made Lift's progress tab
    // unreadable. A week with nothing in it draws no bar, so a gap in a program reads as a gap in the
    // row rather than as a bar the eye has to measure.
    private func weeks(_ weeks: Stats.Weeks) -> some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x4) {
            series(weeks.sets)
            series(weeks.sessions)
            Text(weeks.span)
                .font(GymType.numeral(11))
                .foregroundStyle(skin.inkFaint)
        }
        .padding(WindmillSpace.x4)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(RoundedRectangle(cornerRadius: WindmillRadius.lg).fill(skin.surface))
    }

    private func series(_ series: Stats.Series) -> some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x2) {
            HStack {
                Text(series.label)
                    .font(GymType.numeral(11))
                    .foregroundStyle(skin.inkDim)
                Spacer(minLength: WindmillSpace.x3)
                Text(series.peak)
                    .font(GymType.numeral(11))
                    .foregroundStyle(skin.inkFaint)
            }
            VStack(spacing: 0) {
                HStack(alignment: .bottom, spacing: 3) {
                    ForEach(Array(series.heights.enumerated()), id: \.offset) { _, height in
                        RoundedRectangle(cornerRadius: 1.5)
                            .fill(skin.accent)
                            .frame(height: max(0, height) * 44)
                            .frame(maxWidth: .infinity)
                    }
                }
                .frame(height: 44, alignment: .bottom)
                Rectangle()
                    .fill(skin.line)
                    .frame(height: 1)
            }
        }
        .accessibilityElement(children: .combine)
    }

    // One movement, and the only chart in the product that is a line: a load over sessions is a
    // continuous thing and a count per week is not.
    private func movement(_ line: Stats.Line) -> some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x2) {
            HStack(alignment: .firstTextBaseline) {
                Text(line.movement)
                    .font(WindmillFont.body(17))
                    .foregroundStyle(skin.ink)
                Spacer(minLength: WindmillSpace.x3)
                Text(line.value)
                    .font(GymType.numeral(17, .semibold))
                    .foregroundStyle(skin.ink)
                Text(line.unit)
                    .font(GymType.numeral(11))
                    .foregroundStyle(skin.inkFaint)
            }

            Text(line.lastTrained)
                .font(GymType.numeral(11))
                .foregroundStyle(skin.inkFaint)

            Sparkline(heights: line.heights, ink: skin.accent)
                .frame(height: 36)
                .padding(.vertical, WindmillSpace.x1)

            Text(line.span)
                .font(GymType.numeral(12))
                .foregroundStyle(skin.inkDim)

            ForEach(line.bests, id: \.self) { best in
                Text(best)
                    .font(GymType.numeral(11))
                    .foregroundStyle(skin.inkFaint)
            }
        }
        .padding(WindmillSpace.x4)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(RoundedRectangle(cornerRadius: WindmillRadius.lg).fill(skin.surface))
        .overlay(RoundedRectangle(cornerRadius: WindmillRadius.lg).strokeBorder(skin.line, lineWidth: 1))
    }

    // The honest empty: a log with no FINISHED session has nothing to draw, and the session running
    // right now is deliberately not in here — a statistics read that moved under a lifter between two
    // sets would be reporting on a workout in flight.
    private var nothingYet: some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x2) {
            Text("Nothing finished yet.")
                .font(WindmillFont.body(16))
                .foregroundStyle(skin.inkDim)
            Text("These lines are drawn from sessions you have closed. The one you are in the middle of is today’s screen.")
                .font(GymType.numeral(12))
                .foregroundStyle(skin.inkFaint)
                .lineSpacing(3)
        }
        .padding(WindmillSpace.x4)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(RoundedRectangle(cornerRadius: WindmillRadius.lg).fill(skin.surface))
    }

    // Mono, quiet, never a spinner and never an alert — the room's one voice for a door that did not
    // open. The read is offered again rather than retried forever: this screen is somewhere a lifter
    // chose to stand, and a poll behind it would spend a basement's signal on a chart.
    private func silence(_ line: String, retry: Bool) -> some View {
        VStack(alignment: .leading, spacing: WindmillSpace.x3) {
            Text(line)
                .font(GymType.numeral(13))
                .foregroundStyle(skin.inkFaint)
            if retry {
                Button { Task { await read() } } label: {
                    Text("Try again")
                        .font(WindmillFont.body(16, .semibold))
                        .foregroundStyle(skin.accent)
                        .frame(maxWidth: .infinity, minHeight: GymTap.minimum)
                        .background(RoundedRectangle(cornerRadius: WindmillRadius.lg)
                            .strokeBorder(skin.lineStrong, lineWidth: 1))
                }
            }
        }
    }

    private func read() async {
        failure = nil
        switch await store.statistics() {
        case .success(let statistics):
            board = Stats.board(statistics, catalog: store.catalog,
                                now: Int64(Date().timeIntervalSince1970 * 1000))
        case .failure(let why):
            failure = why
        }
    }
}

// The line itself, and the whole of its drawing: a polyline through the normalised points with the
// newest one marked, because the last point is the one a lifter is looking for. A movement with one
// session is one dot — a chart of a single point is honest and an interpolation to nowhere is not.
private struct Sparkline: View {
    let heights: [Double]
    let ink: Color

    var body: some View {
        GeometryReader { geometry in
            let points = placed(in: geometry.size)
            Path { path in
                for (index, point) in points.enumerated() {
                    index == 0 ? path.move(to: point) : path.addLine(to: point)
                }
            }
            .stroke(ink, style: StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round))

            if let last = points.last {
                Circle()
                    .fill(ink)
                    .frame(width: 5, height: 5)
                    .position(last)
            }
        }
    }

    // Inset by the dot's own radius on every side, so the newest point's mark and a line that runs
    // along its own floor are both drawn INSIDE the frame rather than half over the card's edge.
    private func placed(in size: CGSize) -> [CGPoint] {
        let inset: CGFloat = 3
        let height = size.height - inset * 2
        guard heights.count > 1 else {
            return heights.map { CGPoint(x: size.width / 2, y: inset + height * (1 - $0)) }
        }
        let step = (size.width - inset * 2) / CGFloat(heights.count - 1)
        return heights.enumerated().map { index, value in
            CGPoint(x: inset + CGFloat(index) * step, y: inset + height * (1 - CGFloat(value)))
        }
    }
}

File: apps/android/gym/src/main/kotlin/works/windmill/gym/ui/StatisticsScreen.kt
package works.windmill.gym.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import works.windmill.gym.domain.Stats
import works.windmill.gym.store.GymResult
import works.windmill.gym.store.TrainingStore
import works.windmill.gym.store.WriteFailure
import works.windmill.platform.design.WindmillFont
import works.windmill.platform.design.WindmillRadius
import works.windmill.platform.design.WindmillSpace

// STATISTICS — the long window, and the second place a lifter can stand in this room at rest. It
// RENDERS what `Stats.board` decided and computes nothing itself: no e1RM, no normalisation, no
// sorting, no arithmetic in a drawing.
//
// A CHART ON A PHONE HAS NO SCRUB AND NO HOVER, so every number a mark would otherwise reveal on
// touch is printed beside it — each bar row carries its own peak, each movement line its two
// endpoints. A chart whose values can only be reached by a gesture the surface does not have is a
// decoration, and this product does not draw those.
//
// Nothing here is a grade. No score, no percentage, no green and no red: iris is the one accent
// and it reads as iron, exactly as it does on the weight the lifter is under. Matter-of-fact: no
// entry animation — this is an instrument, not a reveal.
@Composable
fun StatisticsScreen(store: TrainingStore) {
    var board by remember { mutableStateOf<Stats.Board?>(null) }
    var failure by remember { mutableStateOf<WriteFailure?>(null) }
    var asked by remember { mutableIntStateOf(0) }

    // Read on the way in and shaped once, outside a drawing. There is no cache to key on a
    // collection's length and get wrong: a second visit asks the log again, and a set logged in
    // between counts the next time this screen is opened.
    LaunchedEffect(asked) {
        failure = null
        when (val statistics = store.statistics()) {
            is GymResult.Ok -> board = Stats.board(statistics.value, store.catalog, now = System.currentTimeMillis())
            is GymResult.Failed -> failure = statistics.why
        }
    }

    Column(
        verticalArrangement = Arrangement.spacedBy(WindmillSpace.x5),
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = WindmillSpace.x5)
            // Short at the top: this screen is PUSHED, and the frame's own back row sits above it.
            .padding(top = WindmillSpace.x2, bottom = WindmillSpace.x8),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(WindmillSpace.x1)) {
            Text("Statistics", style = WindmillFont.display(32), color = GymSkin.ink)
            Text(
                "every session you have finished",
                style = GymType.numeral(13),
                color = GymSkin.inkFaint,
            )
        }

        val drawn = board
        val silent = failure
        when {
            drawn != null -> {
                drawn.weeks?.let { WeeksCard(it) }
                drawn.lines.forEach { MovementCard(it) }
                if (drawn.isEmpty) NothingYet()
            }
            // In the log's own words when it sent any. A 500 that said "internal error" is not a
            // phone with no signal, and this screen may not report it as one.
            silent != null -> Silence(silent.line("these lines aren’t drawn"), retry = { asked += 1 })
            else -> Silence("reading your log…", retry = null)
        }
    }
}

// Two rows of bars, each measured against ITS OWN peak — sessions and working sets count different
// things, and one axis across both is the chart that made Lift's progress tab unreadable. A week
// with nothing in it draws no bar, so a gap in a program reads as a gap in the row rather than as a
// bar the eye has to measure.
@Composable
private fun WeeksCard(weeks: Stats.Weeks) {
    Column(
        verticalArrangement = Arrangement.spacedBy(WindmillSpace.x4),
        modifier = Modifier
            .fillMaxWidth()
            .background(GymSkin.surface, RoundedCornerShape(WindmillRadius.lg))
            .padding(WindmillSpace.x4),
    ) {
        SeriesRow(weeks.sets)
        SeriesRow(weeks.sessions)
        Text(weeks.span, style = GymType.numeral(11), color = GymSkin.inkFaint)
    }
}

// Bars are counts, measured FROM ZERO against the baseline rule under the row — the domain already
// normalised them that way, and this only spells heights as pixels.
@Composable
private fun SeriesRow(series: Stats.Series) {
    Column(verticalArrangement = Arrangement.spacedBy(WindmillSpace.x2)) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(series.label, style = GymType.numeral(11), color = GymSkin.inkDim)
            Spacer(Modifier.weight(1f))
            Text(series.peak, style = GymType.numeral(11), color = GymSkin.inkFaint)
        }
        Canvas(
            Modifier
                .fillMaxWidth()
                .height(44.dp),
        ) {
            val count = series.heights.size
            if (count == 0) return@Canvas
            val gap = 3.dp.toPx()
            val slot = (size.width - gap * (count - 1)) / count
            series.heights.forEachIndexed { index, height ->
                val tall = (height.coerceIn(0.0, 1.0) * size.height).toFloat()
                if (tall <= 0f) return@forEachIndexed
                drawRoundRect(
                    color = GymSkin.accent,
                    topLeft = Offset(index * (slot + gap), size.height - tall),
                    size = Size(slot, tall),
                    cornerRadius = CornerRadius(1.5.dp.toPx()),
                )
            }
        }
        Box(
            Modifier
                .fillMaxWidth()
                .height(1.dp)
                .background(GymSkin.line),
        )
    }
}

// One movement, and the only chart in the product that is a line: a load over sessions is a
// continuous thing and a count per week is not.
@Composable
private fun MovementCard(line: Stats.Line) {
    Column(
        verticalArrangement = Arrangement.spacedBy(WindmillSpace.x2),
        modifier = Modifier
            .fillMaxWidth()
            .background(GymSkin.surface, RoundedCornerShape(WindmillRadius.lg))
            .border(1.dp, GymSkin.line, RoundedCornerShape(WindmillRadius.lg))
            .padding(WindmillSpace.x4),
    ) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(
                line.movement,
                style = WindmillFont.body(17),
                color = GymSkin.ink,
                modifier = Modifier.alignByBaseline(),
            )
            Spacer(Modifier.weight(1f))
            Text(
                line.value,
                style = GymType.numeral(17, FontWeight.SemiBold),
                color = GymSkin.ink,
                modifier = Modifier.alignByBaseline(),
            )
            Text(
                line.unit,
                style = GymType.numeral(11),
                color = GymSkin.inkFaint,
                modifier = Modifier
                    .alignByBaseline()
                    .padding(start = WindmillSpace.x1),
            )
        }

        Text(line.lastTrained, style = GymType.numeral(11), color = GymSkin.inkFaint)

        Sparkline(line.heights)

        Text(line.span, style = GymType.numeral(12), color = GymSkin.inkDim)

        line.bests.forEach { best ->
            Text(best, style = GymType.numeral(11), color = GymSkin.inkFaint)
        }
    }
}

// The line itself, and the whole of its drawing: a polyline through the normalised points with the
// newest one marked, because the last point is the one a lifter is looking for. A movement with one
// session is one dot — a chart of a single point is honest and an interpolation to nowhere is not.
//
// Inset by the dot's own radius on every side, so the newest point's mark and a line that runs
// along its own floor are both drawn INSIDE the frame rather than half over the card's edge.
@Composable
private fun Sparkline(heights: List<Double>) {
    Canvas(
        Modifier
            .fillMaxWidth()
            .padding(vertical = WindmillSpace.x1)
            .height(36.dp),
    ) {
        if (heights.isEmpty()) return@Canvas
        val inset = 3.dp.toPx()
        val room = size.height - inset * 2
        val points = if (heights.size == 1) {
            listOf(Offset(size.width / 2, inset + room * (1 - heights[0].toFloat())))
        } else {
            val step = (size.width - inset * 2) / (heights.size - 1)
            heights.mapIndexed { index, value ->
                Offset(inset + index * step, inset + room * (1 - value.toFloat()))
            }
        }
        val path = Path()
        points.forEachIndexed { index, point ->
            if (index == 0) path.moveTo(point.x, point.y) else path.lineTo(point.x, point.y)
        }
        drawPath(
            path,
            color = GymSkin.accent,
            style = Stroke(width = 1.5.dp.toPx(), cap = StrokeCap.Round, join = StrokeJoin.Round),
        )
        drawCircle(GymSkin.accent, radius = 2.5.dp.toPx(), center = points.last())
    }
}

// The honest empty: a log with no FINISHED session has nothing to draw, and the session running
// right now is deliberately not in here — a statistics read that moved under a lifter between two
// sets would be reporting on a workout in flight.
@Composable
private fun NothingYet() {
    Column(
        verticalArrangement = Arrangement.spacedBy(WindmillSpace.x2),
        modifier = Modifier
            .fillMaxWidth()
            .background(GymSkin.surface, RoundedCornerShape(WindmillRadius.lg))
            .padding(WindmillSpace.x4),
    ) {
        Text("Nothing finished yet.", style = WindmillFont.body(16), color = GymSkin.inkDim)
        Text(
            "These lines are drawn from sessions you have closed. The one you are in the middle of is today’s screen.",
            style = GymType.numeral(12).copy(lineHeight = 17.sp),
            color = GymSkin.inkFaint,
        )
    }
}

// Mono, quiet, never a spinner and never an alert — the room's one voice for a door that did not
// open. The read is offered again rather than retried forever: this screen is somewhere a lifter
// chose to stand, and a poll behind it would spend a basement's signal on a chart.
@Composable
private fun Silence(line: String, retry: (() -> Unit)?) {
    Column(verticalArrangement = Arrangement.spacedBy(WindmillSpace.x3)) {
        Text(line, style = GymType.numeral(13), color = GymSkin.inkFaint)
        if (retry != null) {
            Box(
                contentAlignment = Alignment.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = GymTap.minimum)
                    .border(1.dp, GymSkin.lineStrong, RoundedCornerShape(WindmillRadius.lg))
                    .clickable(onClick = retry),
            ) {
                Text(
                    "Try again",
                    style = WindmillFont.body(16, FontWeight.SemiBold),
                    color = GymSkin.accent,
                )
            }
        }
    }
}
